src/api/cvc4cpp.cpp

```diff
@@ -2231,6 +2236,207 @@ bool Datatype::const_iterator::operator!=(
   return d_int_ctors != other.d_int_ctors || d_idx != other.d_idx;
 }
 
+/* -------------------------------------------------------------------------- */
+/* Grammar                                                                    */
+/* -------------------------------------------------------------------------- */
+Grammar::Grammar(const Solver* s,
+                 const std::vector<Term>& sygusVars,
+                 const std::vector<Term>& ntSymbols)
+    : d_s(s),
+      d_sygusVars(sygusVars),
+      d_ntSyms(ntSymbols),
+      d_ntsToUnres(),
+      d_dtDecls(),
+      d_allowConst()
+{
+  for (Term ntsymbol : d_ntSyms)
+  {
+    // make the datatype, which encodes terms generated by this non-terminal
+    d_dtDecls.emplace(ntsymbol, DatatypeDecl(d_s, ntsymbol.toString()));
+    // make its unresolved type, used for referencing the final version of
+    // the datatype
+    d_ntsToUnres[ntsymbol] = d_s->getExprManager()->mkSort(ntsymbol.toString());
+  }
+}
+
+void Grammar::addRule(Term ntSymbol, Term rule)
+{
+  CVC4_API_ARG_CHECK_EXPECTED(d_dtDecls.find(ntSymbol) != d_dtDecls.end(),
+                              ntSymbol)
+      << "one of the non terminals given in the predeclaration.";
+
+  CVC4_API_ARG_CHECK_EXPECTED(ntSymbol.getSort() == rule.getSort(), rule)
+      << "the Sorts of rule and ntSymbol to be the same.";
+
+  addSygusConstructorTerm(d_dtDecls[ntSymbol], rule);
+}
+
+void Grammar::addRules(Term ntSymbol, std::vector<Term> rules)
+{
```

Comment:
Check for non-null `ntSymbol`, no non-null terms in `rules`.

Reply:
Done!